src/core/net/srp_client.hpp

```diff
@@ -985,65 +983,113 @@ class Client : public InstanceLocator, private NonCopyable
     };
 #endif // OPENTHREAD_CONFIG_SRP_CLIENT_AUTO_START_API_ENABLE
 
-    struct MsgInfo
+    // SRP Update message encoding methods.
+    class Update
     {
-        static constexpr uint16_t kUnknownOffset = 0;
+    public:
+        Update(const char *aHostName,
+               const char *aDomainName,
+               bool        aServiceKeyRecordEnabled,
+               bool        aUseShortLeaseOption);
+
+        Error    SetMessage(Message *aMessage);
+        Message &GetMessage(void) { return *mMessage; }
+        void     ClearMessage(void);
+        void     ReleaseMessage(void) { mMessage.Release(); }
+        uint16_t GetMessageLength(void) const;
+        void     SetSingleServiceMode(bool aEnabled) { mSingleServiceMode = aEnabled; }
```

Comment:
How about adding methods to set `ServiceKeyRecordEnabled` and `UseShortLeaseOption` instead of passing them as arguments? This would simplify the constructor and avoid the extra `#if OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE` check.

Reply:
Here's another idea that might be simpler:

* We could make `Update` an `InstanceLocator`.
* Then, instead of passing these flags with each `Update` object, they could get the current values by checking with the `Client`. This way, we would only need to track `mServiceKeyRecordEnabled` and `mUseShortLeaseOption` in one place (in `Client`).
* For example, within the `Update` class methods, we could use `Get<Client>().mServiceKeyRecordEnabled`.

What do you think?

Reply:
Using the `InstanceLocator` is simpler. I have updated the code to make `Update` an `InstanceLocator`.